### PR TITLE
cql3: improve expression ergonomics

### PR DESCRIPTION
**Pull request by avikivity**
The `expression` type (an std::variant) suffers from bad ergonomics:
 - std::variant has poor/no constraints, so compiler error messages are long and uninformative
 - it cannot be forward-declared (since std::variant does not support incomplete types)
 - the type name is long, polluting compiler error messages and debug symbols
 - it requires an artificial `nested_expression` when one expression is nested inside another
 
This series fixes those drawbacks by wrapping the variant in a class, adding constraints, and adding an extra indirection.

Test: unit (dev)

### CODE REVIEW

**Review by nyh — APPROVED** [review]
Very nice.

**Review by cvybhu — COMMENTED** [review]
LGTM, I love the readable error messages in `expr::visit`.

I'm a bit worried about having every `expression` in `unique_ptr`, won't that cause a lot of unnecessary allocations?
A collection constructor with a few `ints` inside will now be a `vector<unique_ptr<int32_t>>`.
`nested_expression` was a bit awkward to use, but it wasn't that bad.

We are also gonna need a `expr::visit` that takes a mutable `expression`, but I can add that later when it's needed.

File: cql3/expr/expression.hh [hunk 1]
```diff
@@ -118,26 +129,33 @@ concept invocable_on_expression
...
 
-auto visit(invocable_on_expression auto&& visitor, const expression& expr) {
-    return std::visit(visitor, expr);
-}
+/// A CQL expression -- union of all possible expression types.  bool means a Boolean constant.
+class expression final {
+    // 'impl' holds a variant of all expression types, but since 
+    // variants of incomplete types are not allowed, we forward declare it
+    // here and fully define it later.
+    struct impl;
```
Comment (cvybhu) [inline cql3/expr/expression.hh:137]: There are some trailing spaces here

**Comment by avikivity** [comment]
> LGTM, I love the readable error messages in `expr::visit`.
> 
> I'm a bit worried about having every `expression` in `unique_ptr`, won't that cause a lot of unnecessary allocations?

expressions are only supposed to be manipulated during prepare time. 

> A collection constructor with a few `ints` inside will now be a `vector<unique_ptr<int32_t>>`. `nested_expression` was a bit awkward to use, but it wasn't that bad.
> 

A collection_constructor with a few constants should be folded into a single collection-typed constant!

We could consider using the small-buffer optimization: an expression is 

```c++
class expression {
    impl* _v;  // may point into buf
    char buf[small_expression_size];
};
```

But I don't think it's worth the effort.

> We are also gonna need a `expr::visit` that takes a mutable `expression`, but I can add that later when it's needed.

File: cql3/expr/expression.hh [hunk 2]
```diff
@@ -122,6 +122,23 @@ auto visit(invocable_on_expression auto&& visitor, const expression& expr) {
...
 
+template <ExpressionElement E>
+bool is(const expression& e) {
+    return expr::is<E>(e);
+}
+
+template <ExpressionElement E>
+const E&
+as(const expression& e) {
+    return expr::as<E>(e);
```
[the remark this answers is not in the document]
Comment (avikivity) [reply cql3/expr/expression.hh]: Ugh, it breaks bisection.

[the remark this answers is not in the document]
Comment (avikivity) [reply cql3/expr/expression.hh]: (I used sed to prepare the patch)

**Pull request closed without merging**